### ISSUE TITLE
Fragment setter replaces entire snapshot

### ISSUE THREAD

**Issue by Nuke-** [apollographql/apollo-codegen#240]
The following code is generated using codegen 0.16.5

```swift
public struct SmallImage: GraphQLSelectionSet {
...
    public var url: String? {
        get {
            return snapshot["url"]! as! String?
        }
        set {
            snapshot.updateValue(newValue, forKey: "url")
        }
    }
...
    public var fragments: Fragments {
        get {
            return Fragments(snapshot: snapshot)
        }
        set {
            snapshot = newValue.snapshot
        }
    }
...
```
Here you can see that when a fragment is being set it replaces the entire snapshot.
What should happen is that the snapshot of the new fragment is merged with the existing snapshot.
```swift
snapshot += newValue.snapshot
```
using
```swift
func += <K, V> (left: inout [K:V], right: [K:V]) { 
    for (k, v) in right { 
        left[k] = v
    } 
}
```

**Comment by martijnwalraven** [comment]
Ah, you're right, that makes sense.

I didn't actually consider the case where you would use the setter to replace a fragment with a separate value. In my mind, the setter was there to allow you to change individual properties on the existing snapshot. So you would only use it to 'call through' (like `post.fragments.postDetails.title = "..."`). But once you expose the setter, there is no way to avoid setting the fragment itself of course (`post.fragments.postDetails = ...`).

I'm still trying to figure out what the best way is to handle fragments in the generated code. Out of curiosity, do you have a use case for replacing fragments in this way? Do you feel having fragments as separate values is useful for you?

**Comment by Nuke-** [comment]
In my use case it is convenient to set them separately.
Mainly in unit testing.

with an extensions I added empty constructors to the generated GraphQL objects.
This way I only need to set what I really need in tests and when a new property is being added only the empty constructor needs to be updated in 1 place.

I can then do something as following.

```swift
eventObject = CurrentEventFragment()
eventObject.title = "event_title"
eventObject.endDate = Date().endOfDay
eventObject.startDate = Date().startOfDay

channelInfoObject = ChannelInfoVtvGuide()
channelInfoObject.title = "channel_title"
channelInfoObject.fragments.currentEventFragment = eventObject
```

I could also write it as follow.

```swift
channelInfoObject = ChannelInfoVtvGuide()
channelInfoObject.title = "channel_title"
channelInfoObject.fragments.currentEventFragment.title = "event_title"
channelInfoObject.fragments.currentEventFragment.endDate = Date().endOfDay
channelInfoObject.fragments.currentEventFragment.startDate = Date().startOfDay
```

So no I don't really need it, but I find using individual fragmentObjects is a little more cleaner in unit testing.
And as it is now it might be confusing as it now allows to set the fragment (with unexpected side effect).

**Comment by martijnwalraven** [comment]
@Nuke-: This should be fixed in Apollo iOS 0.7.0-alpha.2, thanks for the report!
